Source/WarpX.cpp

```diff
@@ -340,6 +340,29 @@ WarpX::ReadParameters ()
     {
         ParmParse pp("warpx");
 
+        // set random seed
+        std::string random_seed = "default";
+        pp.query("random_seed", random_seed);
+        if ( random_seed != "default" ) {
+            unsigned long myproc_1 = ParallelDescriptor::MyProc() + 1;
+            unsigned long x = 2;
+            if ( random_seed == "random" ) {
+                unsigned long seed = myproc_1 * std::clock();
+                ResetRandomSeed(seed);
+                AllPrint() << "Actual Random Seed Used: " << seed
+                           << " for MPI process " << myproc_1 - 1 << "\n";
+            } else if ( std::stoul("999"+random_seed) == 999 ) {
+                Abort("Unknown warpx.random_seed value.");
+            } else if ( std::stoi(random_seed) >= 0 ) {
+                unsigned long seed = myproc_1 * ( std::stoul(random_seed) + x );
+                ResetRandomSeed(seed);
+                AllPrint() << "Actual Random Seed Used: " << seed
+                           << " for MPI process " << myproc_1 - 1 << "\n";
+            } else {
+                Abort("Unknown warpx.random_seed value.");
+            }
+        }
```

Comment:
I think we could clarify this piece of code by:
- enforcing n >= 1
- use C++ features for exception handling (what you are doing with `if ( std::stoul("999"+random_seed) == 999 )`)
- removing `AllPrint`, as WarpX runs on tens of thousands of MPI ranks
- Use a random seed that changes in < 1s (if you're submitting automated scans, can take <ms)

Maybe something along these lines? (@ax3l gave some suggestions for this piece of code)

```c++

#include <exception> // at the beginning of the file

          // Set a timer, maybe early in the function
          std::chrono::steady_clock::time_point start = std::chrono::steady_clock::now();
          // set random seed
          int seed_is_specified = pp.query("random_seed", random_seed);
          if ( seed_is_specified ) {
              unsigned long myproc_1 = ParallelDescriptor::MyProc() + 1;
              if ( random_seed == "random" ) {
                  unsigned long seed = myproc_1 * std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() - start).count();;
                  ResetRandomSeed(seed);
              } else {
                  try {
                      unsigned long seed = myproc_1 * ( std::stoul(random_seed) );
                      ResetRandomSeed(seed);
                  } catch (std::invalid_argument my_excepction) {
                      Abort("warpx.random_seed must be \"random\" or a number.");
                  }
              }
          }
```

Reply:
Just remembered: we can also pick a "truly" random number from the entropy pool in C++11 and newer via:
https://en.cppreference.com/w/cpp/numeric/random/random_device